cmd/checkpoint/main.go

```diff
@@ -0,0 +1,268 @@
+)
+
+func main() {
+	log.Println("begin apiserver checkpointing...")
+	run()
+}
+
+func run() {
+	var tempAPIServerManifest v1.Pod
+	if err := json.Unmarshal(manifestBytes, &tempAPIServerManifest); err != nil {
+		log.Fatal(err)
+	}
+	client := newAPIClient()
+	for {
+		var podList v1.PodList
+		if err := json.Unmarshal(getPodsFromKubeletAPI(), &podList); err != nil {
+			log.Fatal(err)
+		}
+		switch {
+		case bothAPIServersRunning(podList):
+			log.Println("both temp and kube apiserver running, removing temp apiserver")
+			// Both the self-hosted API Server and the temp API Server are running.
+			// Remove the temp API Server manifest from the config dir so that the
+			// kubelet will stop it.
+			if err := os.Remove(activeFile); err != nil {
+				log.Println(err)
+			}
+		case kubeSystemAPIServerRunning(podList):
+			log.Println("kube-apiserver found, creating temp-apiserver manifest")
+			// The self-hosted API Server is running. Let's snapshot the pod,
+			// clean it up a bit, and then save it to the ignore path for
+			// later use.
+			tempAPIServerManifest.Spec = parseAPIPodSpec(podList)
+			convertSecretsToVolumeMounts(client, &tempAPIServerManifest.Spec)
+			writeManifest(tempAPIServerManifest)
+			log.Printf("finished creating temp-apiserver manifest at %s\n", ignoreFile)
+
+		default:
+			log.Println("no apiserver running, installing temp apiserver static manifest")
+			b, err := ioutil.ReadFile(ignoreFile)
+			if err != nil {
+				log.Println(err)
+			} else {
+				if err := ioutil.WriteFile(activeFile, b, 0644); err != nil {
+					log.Println(err)
+				}
+			}
+		}
+		time.Sleep(60 * time.Second)
+	}
+}
+
+func getPodsFromKubeletAPI() []byte {
+	var pods []byte
+	res, err := http.Get(kubeletAPIPodsURL)
+	if err != nil {
+		log.Println(err)
+		return pods
+	}
+	pods, err = ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if err != nil {
+		log.Println(err)
+	}
+	return pods
+}
+
+func bothAPIServersRunning(pods v1.PodList) bool {
+	var kubeAPISeen, tempAPISeen bool
+	for _, p := range pods.Items {
+		switch {
+		case isKubeAPI(p):
+			kubeAPISeen = true
+		case isTempAPI(p):
+			tempAPISeen = true
+		}
+		if kubeAPISeen && tempAPISeen {
+			return true
+		}
+	}
+	return false
+}
+
+func kubeSystemAPIServerRunning(pods v1.PodList) bool {
+	for _, p := range pods.Items {
+		if isKubeAPI(p) {
+			return true
+		}
+	}
+	return false
+}
+
+func isKubeAPI(pod v1.Pod) bool {
+	return strings.Contains(pod.Name, "kube-apiserver") && pod.Namespace == api.NamespaceSystem
+}
+
+func isTempAPI(pod v1.Pod) bool {
+	return strings.Contains(pod.Name, "temp-apiserver") && pod.Namespace == api.NamespaceSystem
+}
+
+// cleanVolumes will sanitize the list of volumes and volume mounts
+// to remove the default service account token.
+func cleanVolumes(p *v1.Pod) {
+	volumes := make([]v1.Volume, 0, len(p.Spec.Volumes))
+	for _, v := range p.Spec.Volumes {
+		if !strings.Contains(v.Name, "default") {
```

Comment:
Cannot do exact name match because it's something like `default-token-dsaf8338h` (e.g. default-token with random junk after) IIRC. I can double check that, but maybe it's better to check against `default-token`?